solanaj/src/main/java/org/p2p/solanaj/ws/SubscriptionWebSocketClient.kt

```diff
@@ -0,0 +1,143 @@
+        updateSubscriptions()
+    }
+
+    override fun onMessage(message: String) {
+        Log.d("SOCKET", "New message received: $message")
+        val resultAdapter = Moshi.Builder().build()
+            .adapter<RpcResponse<Long>>(Types.newParameterizedType(RpcResponse::class.java, Long::class.java))
+        try {
+            resultAdapter.fromJson(message)?.also { rpcResult ->
+                val rpcResultId = rpcResult.id
+                if (rpcResultId != null) {
+                    if (subscriptionIds.containsKey(rpcResultId)) {
+                        subscriptionIds[rpcResultId] = rpcResult.result
+                        subscriptions[rpcResultId]?.listener?.let { listener ->
+                            subscriptionListeners[rpcResult.result] = listener
+                        }
+                        subscriptions.remove(rpcResultId)
+                    }
+                } else {
+                    val notificationResultAdapter = Moshi.Builder().build().adapter(RpcNotificationResult::class.java)
+                    notificationResultAdapter.fromJson(message)?.let { result ->
+                        val listener = subscriptionListeners[result.params.subscription]
+                        val value = result.params.result.value as Map<*, *>
+                        when (result.method) {
+                            "signatureNotification" -> listener!!.onNotificationEvent(
+                                SignatureNotification(
+                                    value["err"]
+                                )
+                            )
+                            "accountNotification" -> listener!!.onNotificationEvent(value)
+                        }
+                    }
+                }
+            }
+        } catch (ex: Exception) {
+            Log.e("SOCKET", "Error on socket message", ex)
+        }
+    }
+
+    override fun onClose(code: Int, reason: String, remote: Boolean) {
+        socketStateListener?.onClosed(
+            code,
+            "Connection closed by "
+                + (if (remote) "remote peer" else "us")
+                + " Code: " + code + " Reason: " + reason
+        )
+    }
+
+    override fun onError(ex: Exception) {
+        Log.e("SOCKET", "Error on socket working", ex)
+        socketStateListener?.onFailed(ex)
+    }
+
+    private fun updateSubscriptions() {
+        if (isOpen && subscriptions.isNotEmpty()) {
+            val rpcRequestJsonAdapter = Moshi.Builder().build().adapter(
+                RpcRequest::class.java
+            )
+            for (sub in subscriptions.values) {
+                send(rpcRequestJsonAdapter.toJson(sub.request))
+            }
+        }
+    }
+
+    companion object {
+        private var instance: SubscriptionWebSocketClient? = null
+        private var socketStateListener: SocketStateListener? = null
+        fun getInstance(endpoint: String?, stateListener: SocketStateListener?): SubscriptionWebSocketClient? {
```

Comment:
Singleton pattern can be rewritten using koin::single { }

Reply:
need to make a task for that 
it is out of Scope from this current task